### PR TITLE
[DebugInfo] Salvage debug info for tuples

### PR DESCRIPTION
**Pull request by Snowy1803**
This PR adds support for salvaging removed `tuple` SIL instructions.

- Revert the addition of the `implicit` attribute for debug variables. The implicit attribute meant that implicit variables, such as `self`, `error` in a catch, `newValue` in a property setter, etc., were allowed to be removed at `-Onone`! rdar://125939953
- Add support for salvaging removed `tuple` SIL instructions

With only the second commit, those variables were being removed at Onone and salvaged, which unexpectedly introduced fragments at Onone, which are not compatible with debugger shadow copies. After reverting the implicit attribute, the fragments are now correctly only created when optimizations are enabled.

### CODE REVIEW

**Comment by Snowy1803** [comment]
@swift-ci please test

**Comment by Snowy1803** [comment]
@swift-ci please test windows platform

**Comment by Snowy1803** [comment]
@swift-ci Please Test Source Compatibility

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 1]
```diff
@@ -1875,6 +1871,31 @@ void swift::salvageDebugInfo(SILInstruction *I) {
       }
     }
   }
+  // Similarly, if a `tuple` SIL instruction is "unwrapped" and removed,
+  // we need to make sure any of its related `debug_value` instruction
```
Comment (adrian-prantl) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp:1875]:
```suggestion
  // we need to make sure any of its related `debug_value` instructions
```

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 2]
```diff
@@ -1875,6 +1871,31 @@ void swift::salvageDebugInfo(SILInstruction *I) {
...
+  // is preserved.
```
Comment (adrian-prantl) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp:1876]:
```suggestion
  // are preserved.
```

**Review by adrian-prantl — APPROVED** [review]

**Comment by Snowy1803** [comment]
Source Compatibility failures are the same as on main ([Debug PR](https://ci.swift.org/job/swift-PR-source-compat-suite-debug-macos/1509/) = [Debug on main](https://ci.swift.org/job/swift-main-source-compat-suite-debug/) | [Release](https://ci.swift.org/job/swift-PR-source-compat-suite-macos/1520/) (passing))

**Comment by Snowy1803** [comment]
@swift-ci please smoke test

**Comment by Snowy1803** [comment]
@swift-ci please test windows platform

**Pull request merged**